### ISSUE TITLE
2.4 TODO

### ISSUE THREAD

**Issue by flibitijibibo** [TerryCavanagh/VVVVVV#829]
Wow, that last update didn't take long or anything!

We ended up accumulating a lot of things during the _lengthy_ freeze, so for 2.4 I'm going to intentionally make this a really quick one. 2.4 will only have the following things in it:

- [x] Update to SDL 2.0.20
- [x] Merge/Fix all items labeled 2.4
- [x] Replace SDL_mixer with FAudio
- [x] Everything labeled [localization](https://github.com/TerryCavanagh/VVVVVV/issues?q=is%3Aopen+label%3Alocalization)

If it's a new feature not in the list above, it's probably going to get marked 2.5 unless it's basically written already and is extremely easy to test, because even this list is still a _lot_ of work. Think of 2.4 as mostly a catch-up release to try and undo the damage of the extended freeze. This is also the period where we'll have Misa on as a more official maintainer, so to avoid giving anyone an excessive amount of work right from the get-go I'd like to keep this one MUCH smaller than 2.3.

**Comment by TerryCavanagh** [comment]
Oh, cool, localization! 🎉 

Something long term I'd really like to do is update the mobile versions to be based on the C++ version rather than the adobe air version. Probably that's more 2.5 or 2.6, though, as it's a pretty big undertaking.

**Comment by flibitijibibo** [comment]
A big part of the action set work will be touch integration, so that should be a pretty big step in the right direction. But for things like UI... yeah, it'll take some time :sweat_smile:

**Comment by Fussmatte** [comment]
I'm not gonna lie, I'm SUPER pumped for localisation. It's something I'm passionate about in general, plus it's just super cool to see the game running in other languages. And I love doing localisation myself into Esperanto, like I've done officially for Baba is You and unofficially for some other games and media.

This is all just to say I'm super grateful for all the work people have put in to make this a reality; thank you so much. I'm looking forward to the final release of 2.4!

**Comment by KabanFriends** [comment]
I wonder how localization would be done (like how texts would be translated) for languages like Japanese which was already translated by 3rd-party developers for console releases

Would these text be borrowed from that version, or redone by community or something...?

**Comment by flibitijibibo** [comment]
Did a bunch of work to consolidate the references to SDL_mixer to a single file; functionally it should be identical to before but with most of the audio code now in a single file. This only removed SoundSystem.cpp/h and a single block in main.cpp.

I'll try to get a first draft of the FAudio rewrite this week. It should be easier to test with these changes in place; the only unit that will ever be rebuilt is Music.cpp (unless CMake screws me, which it might?).

**Comment by Fussmatte** [comment]
> I wonder how localization would be done (like how texts would be translated) for languages like Japanese which was already translated by 3rd-party developers for console releases
> 
> Would these text be borrowed from that version, or redone by community or something...?

I apologise for this late reply, and you probably already know this by now but it's likely worth mentioning here so people know.

Terry has mentioned on the Discord server that he's considered contacting Nicalis to use their Japanese localisation as a base point for the PC version. However, it's unlikely that it'll be lifted wholesale to be used for the PC version.

Even if all the text from Nicalis were carried over without changes, the vast amount of new strings that need translation pretty much necessitate a new translation team. I believe the Japanese translators for Dicey Dungeons are who Terry has in mind for this work.

**Comment by flibitijibibo** [comment]
Refactored the TODO a little bit, mostly to separate localization issues from general bugs for 2.4 (which there aren't many of).

Note that the action set stuff has been downgraded quite a bit, it's mostly been boiled down to making _just_ enough of a layer to have keyboard and gamepad bindings that can be used by the localization work. We'll figure out the more robust action set system as part of 2.5.

FAudio is still delayed while I get some other things out the door.

**Comment by flibitijibibo** [comment]
The FAudio port has been merged into master. This went through a _ton_ of iteration and a pretty good amount of testing, but more testing is always welcome - if you're not focused on localization I would focus on testing music and sound effects. (As a reminder, the input work has been demoted to just what's necessary for localization, SDL_ActionSet needs to come first and that's on hold until I can pitch it to a larger sponsor and get the work done).

**Comment by Kokasgui** [comment]
Hey! A question: is the game getting Steam Cloud support in the near future? I think it would greatly benefit it!

**Comment by InfoTeddy** [comment]
> Hey! A question: is the game getting Steam Cloud support in the near future? I think it would greatly benefit it!

It's not on the 2.4 list. But it might be added in a later update. It's been asked for a lot.

**Comment by flibitijibibo** [comment]
All I need to do is set up autocloud, so if it's not set up by the time RC1 is tagged someone should bother me about it.

**Comment by rvbg** [comment]
With the release of the Steam Deck, Steam Cloud support is more important than ever before. Would be awesome to see!

**Comment by flibitijibibo** [comment]
SDL_ActionSet is going to take some more time, so I've cut it from the 2.4 requirements - as soon as localization is ready we can start the process for testing/tagging a new stable release.

**Comment by Dav999-v** [comment]
So here's the status on all remaining [issues tagged either "required for 2.4" or localization](https://github.com/TerryCavanagh/VVVVVV/issues?q=is%3Aopen+is%3Aissue+label%3A%22required+for+2.4%22%2Clocalization)  (excluding the 2.5 ones):
- #936 definitely a must - I've been thinking it'd be good to leave it up to Terry to decide how the screen should look
- #925 yeah this bug should be fixed, this should be pretty easy, just annoying
- #904 I think the end result was that we pretty much decided to at least postpone this for later, and maybe not do it at all
- #866 probably mostly done given the button glyphs work, we just need to lock the editor I think (as discussed in #943)
- #863 this is basically waiting for translations to come in
- #859 this can be closed now, controller layout detection was the last thing that needed to be done

**Comment by flibitijibibo** [comment]
Filed #992 for the editor option. I'm also okay with delaying 904 to 2.5 if everyone else is okay with that.

**Comment by TerryCavanagh** [comment]
Once the last of the localisations are in, I think I'd like to take a fresh look at https://github.com/TerryCavanagh/VVVVVV/issues/904 (probably not) and https://github.com/TerryCavanagh/VVVVVV/issues/860 (maybe) - I think they basically won't happen if they don't happen now, and maybe that's ok, but, you know, let's think it through!

**Comment by flibitijibibo** [comment]
The current build passed Deck cert, so we're in sort of a Last Call mode for 2.4 - anything not immediately essential for 2.4 should probably be bumped to 2.5, which at this point is likely all non-localization tasks and maybe the screenshot and text reload hotkeys. If at all possible it'd be great to have this out next Wednesday, but that's of course an arbitrary deadline - happy to push it back if something urgent comes up at the last minute.

**Comment by Daaaav** [comment]
Yay!

From my perspective nothing's really holding us back, apart from the text reload hotkey!

Another overview of the current "required for 2.4" and "localization" tags, which means we could probably have the update out today or tomorrow if we want:
- #863: only has releasing/publishing steps left
- #904: the decision was made not to hold up 2.4 for it
- #1052: just updates readmes in the repo
- #1053: I guess the maintainer of c-hashmap missed your PR or is busy, maybe all it takes is a followup, but for now the builds could be made with your downstream version, right? (It only affects building with specifically VS2010)
- #1089: we just need to think of an alternative shortcut for reloading language files in translator mode, and update a readme with it
- #1041: I'd be fine not having perfect IME input immediately, the `SDL_HINT_IME_SHOW_UI` in this PR is better than nothing though. We should probably improve it in 2.4.1

Also, there have been developments on supporting an Arabic localization in the past few days! But that should probably also not hold back 2.4 from releasing. The best is probably to add it in a 2.4.x update in a few months, even if 2.5 development has also been started.

**Comment by flibitijibibo** [comment]
Went ahead and bumped c-hashmap to 2.5, since I can use my downstream branch and all it does is fix some keywords for VS2010.

The IME fix also should be okay as-is for 2.4.0, in addition to polishing that feature I may also need to double check something involving text input and key press latency there (the Celeste community found something related to this in FNA recently so I'll have to make sure we're not doing the same thing).

**Comment by TerryCavanagh** [comment]
This is great news :) (I went ahead and merged the repo update issue)

Regarding the Arabic localization - it's a bit of a rush, but the team would like to try it and I think it'd be great to include, so expect some commits around that this week before release! We're now aiming for the 10th!

**Comment by InfoTeddy** [comment]
> Went ahead and bumped c-hashmap to 2.5, since I can use my downstream branch and all it does is fix some keywords for VS2010.

Could the same be done for SheenBidi (which #1093 adds)? In that case, the library has a(n) (easily fixed) Valgrind error (see Tehreer/SheenBidi#19).

Alternatively, I was thinking that I would just fork both repositories with the applied patches and update our submodules to point to those.

**Comment by flibitijibibo** [comment]
We can use downstream submodules for now, yeah - we can try to get back on mainline for both projects after 2.4 is out.

**Comment by flibitijibibo** [comment]
Today's the day! In about 3 hours I'll tag and push 2.4 to `2.4-beta` and send Terry the standalone builds, and Terry will push the novelty oversized launch button an hour after that.

**Comment by TerryCavanagh** [comment]
![giphy](https://github.com/TerryCavanagh/VVVVVV/assets/2874150/ca6b542d-e93e-4aff-9208-4e3ad46ac7f3)

**Comment by flibitijibibo** [comment]
It's live! Also added a 2.3-archive branch on Steam for archiving purposes.

**Comment by flibitijibibo** [comment]
Will likely push a 2.4.1 tomorrow at some point, it's pretty safe as it has mostly language fixes along with a single visual regression fix. #1118 and #1119 also lgtm so if everyone's happy with those MRs I can get those when I tag as well!

**Comment by TerryCavanagh** [comment]
Hey! Actually there are a few more things in the pipeline on the localisation front, including the linegap issue which I want to take a stab at this week. It also doesn't seem like there's anything really critical that needs to be pushed out asap. Can we leave 2.4.1 for two weeks?

**Comment by flibitijibibo** [comment]
Works for me, will hold off on tagging until the localization updates are done.

**Comment by TerryCavanagh** [comment]
Ok! I'd now like to suggest Monday the 5th of February as the release date for 2.4.1. There are a number of localisation update projects happening in the background, but this date should be everyone enough time to get things ready without being rushed.

Hopefully we'll stick on 2.4.1 for a while!

**Comment by flibitijibibo** [comment]
That works for me - SDL 2.30 should be out by that time so the Linux-exclusive controller enhancements will be multiplat for 2.4.1.

**Comment by TerryCavanagh** [comment]
Hey! We should probably give the 2.4.1 update another couple of days to allow for more of a back and forth with the new translators. I think we still need a few adjustments! Shouldn't take too much longer, sorry.

**Comment by flibitijibibo** [comment]
No worries, SDL 2.30 tagged this week so that buys some more time to test changes on my end.

**Comment by flibitijibibo** [comment]
Sounds good, will try and publish on Tuesday!

**Comment by flibitijibibo** [comment]
Will tag in about an hour - at the same time I'll make a `2.4-updates` branch so that the 2.5 work can start without being blocked by bugfix releases.
